File: examples/csharp/extension-geometry-csharp/DcvExtensions/Writer.cs
// /*
//  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
//  * SPDX-License-Identifier: MIT-0
//  *
//  * Permission is hereby granted, free of charge, to any person obtaining a copy of this
//  * software and associated documentation files (the "Software"), to deal in the Software
//  * without restriction, including without limitation the rights to use, copy, modify,
//  * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
//  * permit persons to whom the Software is furnished to do so.
//  *
//  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//  * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//  * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//  */

using System;
using System.IO;
using System.Threading.Tasks;
using Dcv.Extensions;
using Google.Protobuf;

namespace DcvExtensionGeometryCS.DcvExtensions
{
    internal class Writer
    {
        private readonly Stream outputStream = Console.OpenStandardOutput();

        public Writer(Stream outputStream)
        {
            this.outputStream = outputStream;
        }

        public Task SendGetManifestRequest(string requestId)
        {
            return SendRequest(new Request
            {
                RequestId = requestId,
                GetManifestRequest = new GetManifestRequest()
            });
        }

        public Task SendSetupVirtualChannelRequest(string requestId, long pId, string channelName)
        {
            return SendRequest(new Request
            {
                RequestId = requestId,
                SetupVirtualChannelRequest = new SetupVirtualChannelRequest
                {
                    RelayClientProcessId = pId,
                    VirtualChannelName = channelName
                }
            });
        }

        public Task SendCloseVirtualChannelRequest(string requestId, string channelName)
        {
            return SendRequest(new Request
            {
                RequestId = requestId,
                CloseVirtualChannelRequest = new CloseVirtualChannelRequest
                {
                    VirtualChannelName = channelName
                }
            });
        }

        public Task SendSetCursorPointRequest(string requestId, Point point)
        {
            return SendRequest(new Request
            {
                RequestId = requestId,
                SetCursorPointRequest = new SetCursorPointRequest
                {
                    Point = point
                }
            });
        }

        public Task SendGetStreamingViewsRequest(string requestId)
        {
            return SendRequest(new Request
            {
                RequestId = requestId,
                GetStreamingViewsRequest = new GetStreamingViewsRequest()
            });
        }

        public Task SendIsPointInsideStreamingViewsRequest(string requestId, Point point)
        {
            return SendRequest(new Request
            {
                RequestId = requestId,
                IsPointInsideStreamingViewsRequest = new IsPointInsideStreamingViewsRequest
                {
                    Point = point
                }
            });
        }

        private Task SendRequest(Request request)
        {
            return SendMessage(new ExtensionMessage { Request = request });
        }

        private async Task SendMessage<T>(T message) where T : IMessage
        {
            var messageSize = message.CalculateSize();

            var ms = new MemoryStream(messageSize + sizeof(int));
            var bw = new BinaryWriter(ms);

            // Write the size of the serialized protobuf message to the temporary buffer
            bw.Write(messageSize);
            // Write the serialized protobuf message to the temporary buffer
            message.WriteTo(ms);

            var data = ms.ToArray();

            // Send the message (size + serialized protobuf message) to the DCV client
            await outputStream.WriteAsync(data, 0, data.Length);
            await outputStream.FlushAsync();
        }
    }
}

File: examples/rust/extension-virtual-channel-rust/src/writer.rs
// /*
//  * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
//  * SPDX-License-Identifier: MIT-0
//  *
//  * Permission is hereby granted, free of charge, to any person obtaining a copy of this
//  * software and associated documentation files (the "Software"), to deal in the Software
//  * without restriction, including without limitation the rights to use, copy, modify,
//  * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
//  * permit persons to whom the Software is furnished to do so.
//  *
//  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
//  * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
//  * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
//  * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
//  * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//  */
//! This file shows how to write protobuf messages and requests to send to DCV.

use crate::proto::*;
use prost::Message;
use tokio::io::{self, AsyncWriteExt};

// Write a message to send to DCV
async fn write_message(msg: ExtensionMessage) -> io::Result<()> {
    // Use stdout to send messages to DCV
    let mut stdout = io::stdout();

    // Encode the msg size into a 32-bit header in little-endian format
    let header_msg_size = msg.encoded_len();
    stdout
        .write_u32_le(
            header_msg_size
                .try_into()
                .expect("Header(msg size) does not fit into 4 bytes"),
        )
        .await?;
    log::info!(
        "WriteMessage - wrote header(msg_size): {} in stdout handle",
        header_msg_size
    );

    // Encode the message into a buffer
    let mut msg_buffer = vec![];
    msg.encode(&mut msg_buffer).unwrap();
    stdout.write_all(&msg_buffer).await?;
    log::info!("WriteMessage - wrote msg in stdout handle");

    stdout.flush().await?;

    Ok(())
}

// Create a request to send to DCV
pub async fn write_request(request: Request) -> io::Result<()> {
    let ext_msg_request = extension_message::Msg::Request(request);

    let extension_msg = ExtensionMessage {
        msg: Some(ext_msg_request),
    };

    // Send the request as a message to DCV
    let res = write_message(extension_msg).await;
    match &res {
        Ok(()) => log::info!("WriteRequest - Request sent successfully"),
        Err(e) => log::info!("WriteRequest - Request failed. Error: {}", e),
    }

    res
}
